tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/field_relation_scope_test.dart

```diff
@@ -0,0 +1,147 @@
+                ''',
+              )
+              .withFileName('example_class')
+              .build(),
+        ];
+
+        var collector = CodeGenerationCollector();
+        StatefulAnalyzer(config, models, onErrorsCollector(collector))
+            .validateAll();
+
+        expect(collector.errors, isNotEmpty);
+
+        var error = collector.errors.first as SourceSpanSeverityException;
+        expect(error.severity, SourceSpanSeverity.error);
+        expect(
+          error.message,
+          'The relation with scope "serverOnly" requires the relation to be optional.',
+        );
+      },
+    );
+
+    test(
+      'when analyzed then an error is generated for a foreign key relation',
+      () {
+        var models = [
+          parentClassModel,
+          ModelSourceBuilder()
+              .withYaml(
+                '''
+                class: Example
+                table: example
+                fields:
+                  relationId: int
+                  parentClass: ParentClass?, relation(field=relationId), scope=serverOnly
+                ''',
+              )
+              .withFileName('example_class')
+              .build(),
+        ];
+
+        var collector = CodeGenerationCollector();
+        StatefulAnalyzer(config, models, onErrorsCollector(collector))
+            .validateAll();
+
+        expect(collector.errors, isNotEmpty);
+
+        var error = collector.errors.first as SourceSpanSeverityException;
+        expect(error.severity, SourceSpanSeverity.error);
+        expect(
+          error.message,
+          'The relation with scope "serverOnly" requires the relation to be optional.',
+        );
+      },
+    );
+  });
+
+  group('Given a class with an optional relation and "serverOnly" scope', () {
+    test(
+      'when analyzed then no errors are generated for an object relation',
+      () {
+        var models = [
+          parentClassModel,
+          ModelSourceBuilder()
+              .withYaml(
+                '''
+                class: Example
+                table: example
+                fields:
+                  parentClass: ParentClass?, relation(optional), scope=serverOnly
+                ''',
+              )
+              .withFileName('example_class')
+              .build(),
+        ];
+
+        var collector = CodeGenerationCollector();
+        StatefulAnalyzer(config, models, onErrorsCollector(collector))
+            .validateAll();
+
+        expect(collector.errors, isEmpty);
+      },
+    );
+
+    test(
+      'when analyzed then no errors are generated for a foreign key relation',
+      () {
+        var models = [
+          parentClassModel,
+          ModelSourceBuilder()
+              .withYaml(
+                '''
+                class: Example
+                table: example
+                fields:
+                  relationId: int
+                  parentClass: ParentClass?, relation(optional, field=relationId), scope=serverOnly
```

Comment:
Question: is this really possible? Having these two keywords at the same time looks wrong to me, `relation(optional, field=relationId)`. Since `optional` only makes sense if there is no "manual" field defined.

Reply:
`optional` and `field` should probably be mutually exclusive